### PR TITLE
disttask: check affacted row before start subtask

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #49008

Problem Summary:
May have chance that when one task executor want to mark one subtask as running but this subtask is not owned by it.
### What changed and how does it work?
Use the affected rows to check if the executor owns the subtask, if not owned, task executor should not run it.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/disttask/framework/storage/task_table.go [hunk 1]
```diff
@@ -687,35 +671,53 @@ func (stm *TaskManager) HasSubtasksInStates(ctx context.Context, tidbID string,
...
 
 // StartSubtask updates the subtask state to running.
-func (stm *TaskManager) StartSubtask(ctx context.Context, subtaskID int64) error {
-	_, err := stm.executeSQLWithNewSession(ctx, `update mysql.tidb_background_subtask
-		set state = %?, start_time = unix_timestamp(), state_update_time = unix_timestamp()
-		where id = %?`,
-		proto.TaskStateRunning, subtaskID)
-	return err
+func (stm *TaskManager) StartSubtask(ctx context.Context, subtaskID int64, execID string) (bool, error) {
+	// owned means the task executor can own and run the subtask without any other task executor running it.
```
Comment (D3Hunter) [inline pkg/disttask/framework/storage/task_table.go:675]: maybe define a named err as `SubtaskNotFound`？ to mean that the subtask of the id and exec_id is not found

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/disttask/framework/storage/task_table.go (hunk 1 above)
Comment (ywqzzy) [reply pkg/disttask/framework/storage/task_table.go:675]: IMHO, notfound subtask is not an error, if use error may change some error handling code.

Comment (okJiang) [reply pkg/disttask/framework/storage/task_table.go:675]: Notfound errors are very common, I prefer using "error" more. We seem don't need handling this error specifically

image

Comment (ywqzzy) [reply pkg/disttask/framework/storage/task_table.go:675]: > Notfound errors are very common, I prefer using "error" more. We seem don't need handling this error specifically
> 
> image

Indeed we need to filter out NotFound error and do not mark the subtasks on task executor as failed.
See logErrAndPersist

**Review by okJiang — APPROVED** [review]

**Pull request merged**